### PR TITLE
Add range and regexp Intervals

### PR DESCRIPTION
**Pull request by mayya-sharipova**
apache/lucene#13569 introduced IntervalsSource for range and regexp queries. This exposes these features in ES.

This is done to achieve parity with Span queries that support regexp and range.

Relates to #110491

### CODE REVIEW

File: docs/reference/query-dsl/intervals-query.asciidoc [hunk 1]
```diff
@@ -178,6 +180,36 @@ The `pattern` is normalized using the search analyzer from this field, unless
...
+This pattern can expand to match at most 128 terms.
+If the pattern matches more than 128 terms,{es} returns an error.
+
+`pattern`::
+(Required, string) Regexp pattern used to find matching terms.
+For a list of operators supported by the
+`regexp` pattern, see <<regexp-syntax, Regular expression syntax>>.
+
+WARNING: Avoid using wildcard patterns, such as `.*` or `.*?+``. This can
+increase  the iterations needed to find matching terms and slow search
```
Comment (john-wagster) [inline docs/reference/query-dsl/intervals-query.asciidoc:196]: nit: extra space here

**Review by john-wagster — APPROVED** [review]
LGTM

**Review by carlosdelest — APPROVED** [review]

**Review by benwtrent — COMMENTED** [review]
Please don't forget to manually add a changes file to track the feature.

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine test this please

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/packaging-tests-windows-sample

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/windows-2019 / default-windows-archive / packaging-tests-windows-sample

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine test this please

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java [hunk 2]
```diff
@@ -746,6 +752,124 @@ String getUseField() {
...
+            return Objects.hash(pattern, analyzer, useField);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return NAME;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
```
Comment (ChrisHegarty) [inline server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java:824]: @mayya-sharipova while looking over the diff between the `lucene_snapshot` branch and upstream, I noticed this new named writable. Will this need to be accompanied by a new transport version - to indicate the change/addition of the new serialisable type?

**Pull request merged**